File: ios/RadziTracker/TrackingLogger.swift
import Foundation
import os.log

/// Thread-safe ring-buffer log. Snapshots are surfaced to the debug screen via getStatus.
final class TrackingLogger {
  static let shared = TrackingLogger()

  enum Level: String { case info, warn, error }

  struct Entry: Codable {
    let timestamp: Int64
    let level: String
    let message: String
  }

  private let queue = DispatchQueue(label: "com.radzi.tracker.logger")
  private let osLog = OSLog(subsystem: "com.radzi.tracker", category: "tracker")
  private var buffer: [Entry] = []
  private let maxEntries = 500

  private init() {}

  func log(_ level: Level, _ message: String) {
    let entry = Entry(
      timestamp: Int64(Date().timeIntervalSince1970 * 1000),
      level: level.rawValue,
      message: message
    )
    queue.async {
      self.buffer.append(entry)
      if self.buffer.count > self.maxEntries {
        self.buffer.removeFirst(self.buffer.count - self.maxEntries)
      }
    }
    let osType: OSLogType = level == .error ? .error : level == .warn ? .info : .default
    os_log("%{public}@", log: osLog, type: osType, message)
  }

  func snapshot() -> [Entry] {
    queue.sync { self.buffer }
  }

  func clear() {
    queue.async { self.buffer.removeAll() }
  }
}

File: android/app/src/main/java/com/radzi/app/tracking/TrackingLogger.kt
package com.radzi.app.tracking

import android.util.Log

class TrackingLogger internal constructor() {

    enum class Level { info, warn, error }

    data class Entry(val timestamp: Long, val level: String, val message: String)

    private val ring = ArrayDeque<Entry>(CAPACITY)
    private val lock = Any()

    fun log(level: Level, message: String) {
        val entry = Entry(
            timestamp = System.currentTimeMillis(),
            level = level.name,
            message = message
        )
        synchronized(lock) {
            if (ring.size >= CAPACITY) ring.removeFirst()
            ring.addLast(entry)
        }
        when (level) {
            Level.info  -> Log.i(TAG, message)
            Level.warn  -> Log.w(TAG, message)
            Level.error -> Log.e(TAG, message)
        }
    }

    fun snapshot(): List<Entry> = synchronized(lock) { ring.toList() }

    companion object {
        val shared = TrackingLogger()
        private const val CAPACITY = 500
        private const val TAG = "RadziTracker"
    }
}
